pilot/pkg/networking/core/v1alpha3/cluster.go

```diff
@@ -71,20 +75,44 @@ func getDefaultCircuitBreakerThresholds() *cluster.CircuitBreakers_Thresholds {
 // For outbound: Cluster for each service/subset hostname or cidr with SNI set to service hostname
 // Cluster type based on resolution
 // For inbound (sidecar only): Cluster for each inbound endpoint port and for each service port
-func (configgen *ConfigGeneratorImpl) BuildClusters(proxy *model.Proxy, push *model.PushContext) ([]*discovery.Resource, model.XdsLogDetails) {
+// if attemptDeltas is set to true, then BuildClusters will attempt to build delta clusters for that proxy
+func (configgen *ConfigGeneratorImpl) BuildClusters(proxy *model.Proxy, push *model.PushContext, attemptDeltas bool, updates *model.PushRequest, watched *model.WatchedResource) ([]*discovery.Resource, model.XdsLogDetails) {
 	clusters := make([]*cluster.Cluster, 0)
 	resources := model.Resources{}
+	services := make([]*model.Service, 0)
 	envoyFilterPatches := push.EnvoyFilters(proxy)
 	cb := NewClusterBuilder(proxy, push, configgen.Cache)
 	instances := proxy.ServiceInstances
-	cacheStats := cacheStats{}
+	// delta when only services are modified and we attempt deltas
+	delta := updates != nil && allConfigKeysOfType(configMapKeys(updates.ConfigsUpdated), gvk.Service) && attemptDeltas
+	if delta {
+		// use delta -- only services changed
+		configs := model.ConfigsOfKind(updates.ConfigsUpdated, gvk.Service)
+		for s := range configs {
+			// get service that changed
+			service := push.ServiceForHostname(proxy, host.Name(s.Name))
+			// is the service visible to envoy?
+			// does envoy care about this service changing? (watchedResources)
+			if proxy.ServiceInSidecarScope(service) {
+				services = append(services, service)
+			}
+		}
+	}
+	clusterCacheStats := cacheStats{}
 	switch proxy.Type {
 	case model.SidecarProxy:
 		// Setup outbound clusters
 		outboundPatcher := clusterPatcher{efw: envoyFilterPatches, pctx: networking.EnvoyFilter_SIDECAR_OUTBOUND}
-		ob, cs := configgen.buildOutboundClusters(cb, outboundPatcher)
-		cacheStats = cacheStats.merge(cs)
+		var ob []*discovery.Resource
+		var cs cacheStats
+		if delta {
+			ob, cs = configgen.buildOutboundClustersWithServices(cb, outboundPatcher, services)
```

Comment:
I wonder if this code would be a bit simpler if we just always use a single buildOutboundClusters call, but allow passing nil `services` to generate all?

Although need to be careful to not accidentally send all if the delta logic returns no services, so maybe too tricky